services/matching-service/start.ts

```diff
@@ -0,0 +1,120 @@
+import express from "express";
+import { Server } from "socket.io";
+import { createServer } from "http";
+
+import { extractCategory, extractDifficulty, isDuplicateID as checkIdExists } from "./domain/match_validators";
+
+const app = express();
+const httpServer = createServer(app);
+const io = new Server(httpServer);
+
+
+const port = process.env.PORT || 6001;
+
+io.on("connection", (socket) => {
+    socket.emit("reply", "hello from server");
+
+    socket.on("register", (msg) => {
+        console.log(`A NEW USER WITH ID ${msg.id} JOINED!`);    
+    
+        const difficulty = extractDifficulty(msg.difficulty);
+        const category = extractCategory(msg.category);
+        
+        // Check Valid Difficulty
+        if (difficulty == undefined) {
+            socket.emit("error", "Invalid Difficulty!");
+            return;
+        }
+        
+        //Check valid Category
+        if (category == undefined) {
+            socket.emit("error", "Invalid Category!");
+            return;
+        }
+        
+        // Check for duplicate ID and push to Queue
+        if (checkIdExists(msg.id)) {
+            socket.emit("error", "You are already in the queue!");
+            return;
+        } 
+
+        // Find match if exists, else create new queue and wait.
+        const this_match = {
+                id: msg.id,
+                difficulty: difficulty,
+                category: category,
+                sockAddr: socket.id
+            }
+        const queueName = difficulty + category; // e.g. easyrecursion
+        let this_queue = queues.get(queueName);
+        if (this_queue) {
+            // Match found, remove queue from queues
+            this_queue.push(this_match)
+            console.log("WE FOUND A MATCH!");
+            queues.delete(queueName);
+            
+            // Remove id from ids.
+            const peerID = this_queue[0].id;
+            ids = ids.filter(id => id != msg.id && id != peerID);
+            
+            // Emit on this user's socket
+            socket.emit("success", `You have been paired with User ${peerID}.`);
```

Comment:
Maybe we could check if peerId is still connected (like what you did in line 65/69) before returning the id, else if the they had already disconnected, our user would be stuck in an empty room